### PR TITLE
Show Uncertainty Glyphs

### PR DESCRIPTION
**Pull request by RubioJr9**
-Added module to visualize tensor field uncertainty using glyphs
-Added uncertainty glyph builder class to tensor glyph builder file
-Added test network
-Moved some files from Core::Algorithm that should be in Graphics::Datatypes

Note: The glyph surface is not mathematically correct, so I put an experimental label in the module

### CODE REVIEW

**Comment by dcwhite** [comment]
```
/Users/runner/work/SCIRun/SCIRun/src/Interface/Modules/Visualization/ShowUncertaintyGlyphsDialog.cc:37:77: error: unused parameter 'name' [-Werror,-Wunused-parameter]
ShowUncertaintyGlyphsDialog::ShowUncertaintyGlyphsDialog(const std::string& name,
```

File: src/Graphics/Glyphs/TensorGlyphBuilder.cc [hunk 1]
```diff
@@ -37,6 +37,213 @@ using namespace Core::Geometry;
...
+  const Eigen::Matrix<double, 6, 6>& covarianceMatrix)
+    : TensorGlyphBuilder(t, center), emphasis_(emphasis), covarianceMatrix_(covarianceMatrix)
+{
+  h_ = 0.000001;// * t.frobeniusNorm();
+  hHalf_ = h_ * 0.5;
+}
+
+void UncertaintyTensorOffsetSurfaceBuilder::generateOffsetSurface(GlyphConstructor& constructor)
+{
+  Vector centerVector = Vector(center_);
```
Comment (dcwhite) [inline src/Graphics/Glyphs/TensorGlyphBuilder.cc:51]:
```suggestion
  Vector centerVector(center_);
```

File: src/Modules/Visualization/ShowUncertaintyGlyphs.cc [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+ShowUncertaintyGlyphsImpl& ShowUncertaintyGlyphsImpl::setTensorsScale(double scale)
+{
+  scale_ = scale;
+  return *this;
+}
+
+
+// TODO move
+enum FieldDataType
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowUncertaintyGlyphs.cc:149]: There's already an enum for this, can you use that one?

File: src/Graphics/Glyphs/TensorGlyphBuilder.h [hunk 3]
```diff
@@ -97,6 +110,49 @@ class SCISHARE TensorGlyphBuilder
   int nv_ = 0;
   int nu_ = 0;
 };
+
+  using MandelVector = Eigen::Matrix<double, 6, 1>;
+
+  struct DifftValues
```
Comment (dcwhite) [inline src/Graphics/Glyphs/TensorGlyphBuilder.h:116]: Not a fan of a big struct of public arrays, with a private magic number size. I'll look at the usage and try to suggest something.

**Review by dcwhite — CHANGES_REQUESTED** [review]

File: src/Modules/Visualization/ShowUncertaintyGlyphs.cc (hunk 2 above)
Comment (RubioJr9) [reply src/Modules/Visualization/ShowUncertaintyGlyphs.cc:149]: It's in 2 other modules. I was planning on moving this, just don't know where to put it

File: src/Graphics/Datatypes/RenderFieldState.h [hunk 4]
```diff
@@ -26,12 +26,11 @@
...
 
-#ifndef CORE_ALGORITHMS_VISUALIZATION_RENDER_FIELD_STATE_H
-#define CORE_ALGORITHMS_VISUALIZATION_RENDER_FIELD_STATE_H
+#ifndef GRAPHICS_DATATYPES_RENDER_FIELD_STATE_H
+#define GRAPHICS_DATATYPES_RENDER_FIELD_STATE_H
 
 #include <Core/Datatypes/Color.h>
-#include <array>
-#include <Core/Algorithms/Visualization/share.h>
+#include <Graphics/Datatypes/share.h>
```
Comment (dcwhite) [inline src/Graphics/Datatypes/RenderFieldState.h:33]:
```suggestion
#include <array>
#include <Graphics/Datatypes/share.h>
```

Also needed for Windows

File: src/Modules/Visualization/ShowUncertaintyGlyphs.cc (hunk 2 above)
Comment (dcwhite) [reply src/Modules/Visualization/ShowUncertaintyGlyphs.cc:149]: You can reuse `Mesh::NODES_E, Mesh::FACES_E`, and the like instead.

**Comment by RubioJr9** [comment]
I'm going to add comments to explain the uncertainty glyph's variable naming soon. I wanted to get the builds going first.

**Review by jessdtate — COMMENTED** [review]
It seems that the major "un-named" variables are qn and P.  qn seems to be a discretization of q or Q from the paper.    Does that match what see @RubioJr9?  if so, maybe it would be better as q_i to be more consistent with the paper?  p is a point/angle on the glyph, right?  I think it's fine to leave it as p in that case.  

I didn't see others that weren't part of loops.

**Comment by RubioJr9** [comment]
> It seems that the major "un-named" variables are qn and P. qn seems to be a discretization of q or Q from the paper. Does that match what see @RubioJr9? if so, maybe it would be better as q_i to be more consistent with the paper? p is a point/angle on the glyph, right? I think it's fine to leave it as p in that case.
> 
> I didn't see others that weren't part of loops.

I don't think it is a discretization. I think the qn variable is a measure of finite differences in each of the 6 degrees of freedom of the mean tensor.

**Comment by dcwhite** [comment]
@jessdtate Just waiting on your approval on this one too

**Review by dcwhite — APPROVED** [review]

**Review by jessdtate — APPROVED** [review]

**Pull request merged**